usr/lib/api/socket_client.c

```diff
@@ -23,114 +23,423 @@
         if (n < 0) {
             // read error
             if (errno == EINTR)
                 continue;
-            OCK_SYSLOG(LOG_ERR, "init_socket_data: read error \
-                       on daemon socket, errno=%d", errno);
-            goto exit;
-        } else if (n == 0) {
-            // eof but we still expect some bytes
-            OCK_SYSLOG(LOG_ERR, "init_socket_data: read returned \
-                       with eof but we still \
-                       expect %lu bytes from daemon",
-                       sizeof(*daemon_socket_data) - bytes_received);
-            goto exit;
-        } else {
-            // n > 0, we got some bytes
-            bytes_received += n;
+            return -errno;
         }
+        if (n == 0)
+            break;
+
+        bytes_received += n;
     }
 
-    ret = TRUE;
+    return bytes_received;
+}
+
+static ssize_t send_all(int socketfd, char *buffer, size_t size)
+{
+    size_t bytes_sent = 0;
+    ssize_t n;
 
-    // copy the Slot_Mgr_Socket_t struct into global
-    // Anchor SocketDataPdata buffer
-    memcpy(&(Anchor->SocketDataP), daemon_socket_data,
-           sizeof(*daemon_socket_data));
+    while (bytes_sent < size) {
+        n = send(socketfd, buffer + bytes_sent, size - bytes_sent, 0);
+        if (n < 0) {
+            // send error
+            if (errno == EINTR)
+                continue;
+            return -errno;
+        }
+        if (n == 0)
+            break;
 
-exit:
-    //free the data buffer after copy
-    if (daemon_socket_data)
-        free(daemon_socket_data);
+        bytes_sent += n;
+    }
 
-    close(socketfd);
+    return bytes_sent;
+}
+
+//
+// Will fill out the Slot_Mgr_Socket_t structure in the Anchor global data
+// structure with the values passed by the pkcsslotd via a socket RPC.
+int init_socket_data(int socketfd)
+{
+    ssize_t n;
+    int ret = TRUE;
+
+    n = read_all(socketfd, (char *)&Anchor->SocketDataP,
+                 sizeof(Anchor->SocketDataP));
+    if (n < 0) {
+        // read error
+        OCK_SYSLOG(LOG_ERR, "init_socket_data: read error \
+                   on daemon socket, errno=%d", -n);
+        ret = FALSE;
+    }
+    if (n != sizeof(Anchor->SocketDataP)) {
+        // eof but we still expect some bytes
+        OCK_SYSLOG(LOG_ERR, "init_socket_data: read returned \
+                   with eof but we still \
+                   expect %lu bytes from daemon",
+                   sizeof(Anchor->SocketDataP) - n);
+        ret = FALSE;
+    }
 
     return ret;
 }
+
+static bool match_token_label_filter(event_msg_t *event, API_Slot_t *sltp)
+{
+    if (event->token_label[0] == ' ' || event->token_label[0] == '\0')
+        return true;
+
+    return memcmp(event->token_label,
+                  sltp->TokData->nv_token_data->token_info.label,
+                  sizeof(event->token_label)) == 0;
+}
+
+#define member_size(type, member) sizeof(((type *)0)->member)
```

Comment:
This definition appears at leat 2 times. Is there a suitable header for those kind of macros ?

Reply:
Well, actually the one definition in events.h should be sufficient, that include file is included in all the others that need that definition. 

Let me change that.